### ISSUE TITLE
Database independant, Table specific DAOs?

### ISSUE THREAD

**Issue by clragon** [simolus3/moor#2562]
**Is your feature request related to a problem? Please describe.**
I would like to make modular code. Some operations are more object-specific (table) than database specific.
I would like to be able to write code specific to a table and then reuse it in different database definitions.

**Describe the solution you'd like**
I would like to create DAOs which database independant. DAOs already specify which tables they use and I might want to reuse my DAO on different database definitions.

However, `DatabaseAccessor` requires a `GeneratedDatabase` so right now it doesnt seem possible to pull this off.
The generated Table definitions only seem to exist if a database is actually present; and they are not public either.
So its not really possible to "reuse" them. I also have not seen any type to say "the generated version of this Table type".

**Comment by simolus3** [comment]
You can do this by enabling drift's [modular code generation](https://drift.simonbinder.eu/docs/advanced-features/builder_options/#modular-code-generation). It requires a little bit of additional setup around managing imports, so it's not enabled by default. But it should work great when you're using the same table definition with multiple database classes.

When enabled, drift emits an independent Dart file for each Dart or `.drift` input file that defines a table or something else understood by drift. Multiple databases can then use the same table classes without drift generating duplicate table definitions. You can also use these tables in shared, table-specific DAOs.

There's an example for modular code generation [in this repository](https://github.com/simolus3/drift/tree/develop/examples/modular). It uses drift files and no database-agnostic DAOs, but Dart-defined tables are supported as well.

To make a DAO support multiple database class, you can define it with `GeneratedDatabase`:

```dart
@DriftAccessor(
  include: {'src/user_queries.drift'},
  queries: {'addUser': r'INSERT INTO users $user;'},
)
class MyAccessor extends DatabaseAccessor<GeneratedDatabase>
    with $MyAccessorMixin {
  MyAccessor(super.attachedDatabase);
}
```

`GeneratedDatabase` is the superclass of all drift databases, so this DAO can be added to every database class you'll write.

**Comment by clragon** [comment]
Okay, that seems to work. kind of.

I was trying to achieve something like this:

```dart
@UseRowClass(Item, generateInsertable: true)
class ItemTable extends Table {
  IntColumn get id => integer().autoIncrement()();
}

@DriftAccessor(tables: [ItemTable])
class ItemDao extends DatabaseAccessor<GeneratedDatabase> {
  Stream<Item> get(int id) =>
      (select(itemsTable)..where((tbl) => tbl.id.equals(id)))
          .watchSingle();
}
```

and if I enable modular code generation and use an import instead of a `part of`, this does indeed work. 

I do run into some issues though;

- The generated code seems to turn out incorrectly. 
  when using a CustomRowClass with `generateInsertable: true`, the generated code doesnt use the import prefixes, so it fails to compile
- The toString method seems to also use the import prefix on the companion, which results in a very odd looking output
- This breaks all my other definitions, because its a global setting
- I now have to make this generated `whatever.drift.dart` file a public thing or create extra classes to export it
  I prefer using part files that generate mixins for my classes so I can then export my classes, hiding away all the generated code. 
  the names for generated classes arent very nice and I would like to shield the rest of my app from it. 
  (Maybe this could be solved by me explicitly defining a class for the generated Table which then gets a mixin in the part file?) (How would this be different from just making it now with the public class? I suppose it wouldnt break all my other definitions which arent using this way of doing databases) (But, drift then wouldnt know which Table to use; conceptually, if the Table and the generated Table would be the same class from the beginning, this could work out. It would also mean my DAO wouldnt have to be code generated. I could simply say "I use this (generated) table". That sounds like a very large breaking everything change. But I think it might be more "modular" by default, which a much lower entry cost (0)?)

I understand that some of the problems I have brought up there are probably very complex and have far reaching consequences if attempted to solve, especially the last one.

I am very sorry for my rather unorganized thoughts there. If you would like I can make a separate issue for those things.

**Comment by simolus3** [comment]
So the first two points are bugs in the implementation obviously, I've fixed them in 9c21e17d912ab2b9fca5392a2ffb3a14cef89067.

> This breaks all my other definitions, because its a global setting

The documentation about this isn't obvious, but you can split your build into [build targets](https://pub.dev/packages/build_config) and apply different drift options to different parts of your project. You can see an example for that [here](https://github.com/simolus3/drift/blob/develop/drift/build.yaml).

> file a public thing or create extra classes to export it

This only affects companions, right? If I'm not missing something, these `.drift.dart` files should otherwise only be imported by other `.drift.dart` files, and drift manages those imports for you. You just import your regular table definitions.

One reason we're not making them part files is because they might need stuff the main library doesn't have. I think it can't happen with Dart files alone, but if you do something like

```dart
import 'item_table.dart';

@DriftAccessor(tables: [ItemTable], queries: {'foo': 'SELECT some_column FROM items'})
```

And `item_table.dart` imports a type converter applied on `some_column`, the generated code for the accessor breaks because the types for that type converter have not been imported into the Dart file defining the accessor. This might sound like a constructed example, but it has been an annoyance for users of the default builder. By letting the generator manage imports, we can avoid this problem.

> But I think it might be more "modular" by default, which a much lower entry cost (0)?)

Well, the entry cost is applying a generated mixin to every table, right?

**Comment by clragon** [comment]
> The documentation about this isn't obvious, but you can split your build into [build targets](https://pub.dev/packages/build_config) and apply different drift options to different parts of your project. You can see an example for that [here](https://github.com/simolus3/drift/blob/develop/drift/build.yaml).

Ah I see. 

> This only affects companions, right? If I'm not missing something,

Right, I was thinking manually a bit too much. I would have kinda liked the idea of just generating a Table somewhere and then using it elsewhere without having to have a intricate project wide link up through a code generator. This works for me though.

> One reason we're not making them part files is because they might need stuff the main library doesn't have.

I am going to be honest I can't follow. I have no idea how that works and I avoid raw SQL whereever I can. But if you say so then I take it as that.

> Well, the entry cost is applying a generated mixin to every table, right?

Thats true. I think if it was optional then I would prefer it over writing my `build.yaml` in what seems to be a very complicated setup. Maybe modular code generation could be available as a annotation as well? (Though I am not sure which class would be annotated with that).
